crates/bevy_animation/src/blend.rs

```diff
@@ -0,0 +1,189 @@
+};
+use bevy_math::*;
+use bevy_transform::components::Transform;
+
+/// A type with a natural method of addative blending.
+///
+/// Any type with a well-behaved method of composition which also implements the [`Interpolate`] trait
+/// can support addative blending. Specifically, the type must:
+///
+/// 1. Have a natural method of composition. For vectors this is addition, for quaternions or rotations
+///    composition is usually expressed as multiplication. We'll use `comp(a, b)` to denote this operation.
+///
+/// 2. The operation must be associative. That is, `comp(comp(a, b), c)` must be equivalent to
+///    `comp(a, comp(b, c))`. The result does not need to be data-identical, but it should be equivalent
+///    under some reasonable notion of equivalence.
+///
+/// 2. Have an `IDENTITY` value such that composition with the identity is equivalent in some way to the
+///    original value.
+///
+/// 3. The value of `T::blend(a, b, w)` should be equivalent to `comp(a, interp(IDENTITY, b, w))`. This implies
+///    that `T::blend(a, b, 0)` is equivalent to `a` and `T::blend(a, b, 1)` is equivalent to `comp(a, b)`.
+///
+/// Some of you will have noticed that these rules encodes the axioms for a Monoid; In fact this trait
+/// represents something similar to a Lie Group.
+pub trait Blend: Interpolate {
+    /// The default value of the blend, which has no effect when blended with other values.
+    const IDENTITY: Self;
+
+    /// Addatively blends another value on top of this one.
+    fn blend(self, other: Self, blend_weight: f32) -> Self;
+}
+
+/// The state for an ongoing blend operation. On types implementing `Blend`, you can start a new
+/// blend operation using [`Blendable::blend_additive`] or [`Blendable::blend_interp`].
+pub struct Blender<T>
+where
+    T: Blend,
+{
+    /// The value being blended.
+    value: T,
+    /// The cumulative weight of the blend operation so far. Every new blend operation
+    /// increases this weight.
+    weight: f32,
+}
+
+impl<T> Blender<T>
+where
+    T: Blend,
+{
+    /// Addatively blends the value into the blender.
+    pub fn blend_additive(self, value: T, weight: f32) -> Self {
+        Blender {
+            value: T::blend(self.value, value, weight),
+            weight: self.weight + weight,
+        }
+    }
+
+    /// Interpolatively blends the value into the blender.
+    pub fn blend_interp(self, value: T, weight: f32) -> Self {
+        let cumulative_weight = self.weight + weight;
+        Blender {
+            value: T::interp(&self.value, &value, weight / cumulative_weight),
+            weight: cumulative_weight,
+        }
+    }
+
+    /// Finishes the blend and returns the blended value.
+    pub fn finish(self) -> T {
+        self.value
+    }
+}
+
+/// This is an extension trait to `Blend` with methods to easily blend and interpolate
+/// between multiple values.
+pub trait Blendable: Blend {
+    /// Addatively blends the value into a new blender.
```

Comment:
```suggestion
    /// Additively blends the value into a new blender.
```

Reply:
Would you believe I put this through two different levels of spell-check and these still got through somehow?